### DISCUSSION TITLE
VSCode autocompletion

### DISCUSSION THREAD

**Discussion by idogrf** [custom-components/pyscript#399] [Q&A]
Hi,

Was wandering if anyone knows of a way for getting autocompletions for pyscript's entities in vscode in .py files 
for example, when you write input_boolean - get all the possible options

I know this is possible using jupyter, and I thought I could use vscode's jupyter extension instead of the Jupyter Notebook UI...
I can start the jupyter server and connect to it using vscode, I can even run things, by I don't have autocompletion here either

any ideas?

**Comment by dingedan** [comment]
I just started working on a script that generates a [builtins.pyi](https://github.com/microsoft/pyright/blob/main/docs/builtins.md) file for pylance to accomplish exactly this. I am a total beginner to both python and visual studio code so I don't know if this is best way but it's working pretty well. Right now the script runs on my dev client and i manually run it when my home assistant setup changes but it should be possible to run it in the Pyscript server environment so that it automatically keeps up to date. I will probably be able to share in a while.

**Reply by contagon** [reply]
Been looking into this myself, did you ever get very far with it? This is something that'd really help with developing pyscript automations

**Comment by Morriz** [comment]
Can we broaden the scope and also get the basics working? I now am not able to get autocompletion on the `hass` object. It is suggested to configure as a global, which kills autocompletion, and there is NO documentation on how to import it.

**Comment by redlefloh** [comment]
I am just repinging this thread since I started using VS Code for pyscript development and especially now with Co-Pilot it is so much better compared to developing in a Jupyter notebook. I did get the remote hass kernel configured and running in VS Code. However, what is **really** missing is IntelliSense for e.g. entities, functions, etc. coming from the hass kernel. The Jupyter Notebook supports that, but unfortunately there does not seem to be a solution for VS Code, yet? Does anyone have any pointers?

**Comment by ALERTua** [comment]
Code completion is now out of the box: https://github.com/custom-components/pyscript/pull/771

**Reply by redlefloh** [reply]
OMG, brilliant, can;t wait for it to surface in HACS!!!

**Comment by redlefloh** [comment]
Thanks so much @dmamelin. I installed the master branch to test it but noticed that [the documentation](https://hacs-pyscript.readthedocs.io/en/latest/reference.html#ide-helpers:~:text=VS%20Code%3A%20open%20Settings%20%E2%86%92%20Python%20Analysis%3A%20Extra%20Paths%20%2D%3E%20Add%20Item%20%2D%3E%20pyscript/modules.) seems outdated, it says:
`VS Code: open Settings → Python Analysis: Extra Paths -> Add Item -> pyscript/modules.`
but shouldn't it be:
`VS Code: open Settings → Python Analysis: Extra Paths -> Add Item -> pyscript/stubs.`
i.e. in .vscode/settings.json it should be:
```
  "python.analysis.extraPaths": [
    "./custom_components/pyscript/stubs"
  ]
```
Correct?

But even with that setting, autocompletion is not yet triggereing for me? The hass kernel is running fine. 

Am I missing something @dmamelin? Thanks!

**Reply by dmamelin** [reply]
The docs are correct.
`pyscript.generate_stubs` writes files to pyscript/modules/stubs/.
In your IDE, add pyscript/modules as a source root.

Did you run `pyscript.generate_stubs`?

**Reply by redlefloh** [reply]
oh, my bad, thanks! Yes, I did run `pyscript.generate_stubs`, but I was looking for the generated files under `custom_components/pyscript/stubs` and I didn't look sharp enough but only saw there was no modules folder, so I thought it was a mistake in the documentation and added the folder `./custom_components/pyscript/stubs` to the settings, without modules but with ./custom_components/. That obviously couldn't work.

I now realized that the generated files are actually in the folder `pyscript/modules/stubs`. I therefore changed the settings as described in the documentation to `pyscript/modules` but unfortunately after a `Developer Reload Window` I still don't get the autocompletion to work. 

Here is again what I did:
- I am running Home Assistant in a docker container on a RasPi 5

- I am also running a jupter server on the RasPi 5

- I run VSCode on a Chromebook and use remote ssh to access the folders and connect to the jupyter server on the RasPi 5 

- I installed the latest master branch for pyscript with your stub generator addition on the RasPi, restarted Home Assistant, everything works fine 

- I run `pyscript.generate_stubs` in Home Assistant and found the two generated files in `pyscript/modules/stubs` 

- I added `pyscript/modules` to the "python.analysis.extraPaths" settings in Workspace:
<img width="452" height="130" alt="image" src="https://github.com/user-attachments/assets/ef7a404e-883c-40f1-91c2-a5e5fef883ce" />

- and I also added `pyscript/modules` to the "python.analysis.extraPaths" settings in User (not strictly needed but just in case):
<img width="452" height="130" alt="image" src="https://github.com/user-attachments/assets/15b1c2f6-2b11-4721-b3a7-c8b9703b770f" />

- I restarted VS Code

- I openend a Notebook, typed `sensor.` in a cell and Ctrl-Space, it briefly shows '"Loading"' but then '"No Suggestions"':

<img width="452" height="83" alt="image" src="https://github.com/user-attachments/assets/632d9119-4a50-406f-bb25-07be0a36a7a5" />

- I also see this in the Output:
<img width="690" height="94" alt="image" src="https://github.com/user-attachments/assets/23f74970-2800-4015-a4f1-2d3234e48d75" />

In addition to `pyscript/modules` in the "python.analysis.extraPaths" settings, I also tried: 
- `./pyscript/modules` 
- `pyscript/modules/stubs`
- `./pyscript/modules/stubs`

but no success?

I am not sure what I am still missing? It sounded so simple and straightforward and I can't wait to be able to use it.
Any other thought? Thanks so much!

**Reply by dmamelin** [reply]
`from stubs.pyscript_builtins import sensor` or `from stubs.pyscript_builtins import *`

**Reply by redlefloh** [reply]
oh no, of course, too many trees to see the forest. Works like a charm. In Jupyter Notebooks I never needed to import anything so I was just not thinking... Thanks again for the amazing work @dmamelin and for the help on my last inches!
